examples/common.c

```diff
@@ -391,3 +391,58 @@ int cred_acquire_cb(git_cred **out,
 	free(pubkey);
 	return error;
 }
+
+char *read_file(const char *path) {
+	FILE *file;
```

Comment:
May I propose this alternate implementation:

```
char *read_file(const char *path)
{
    ssize_t total = 0;
    char *buf = NULL;
    struct stat st;
    int fd = -1;

    if ((fd = open(path, O_RDONLY)) < 0 || fstat(fd, &st) < 0)
        goto out;

    if ((buf = malloc(st.st_size + 1)) == NULL)
        goto out;

    while (total < st.st_size) {
        ssize_t bytes = read(fd, buf + total, st.st_size - total);
        if (bytes <= 0 && errno != EAGAIN && errno != EINTR) {
            free(buf);
            buf = NULL;
            goto out;
        }
        total += bytes;
    }

    buf[total] = '\0';

out:
    if (fd >= 0)
        close(fd);
    return buf;
}
```

Reply:
Would that work on Windows?

Reply:
Fair question, but yes it does, I've just tested with MSVC2017. It requires some different header files than on Unix systems, though, most importantly "io.h"

Reply:
Done

Reply:
This code breaks on linux:

```
$ gdb --args ./examples/lg2 diff --no-index --cached examples/file1 examples/file2
GNU gdb (GDB) 8.3
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./examples/lg2...
(gdb) run
Starting program: /home/alberto/projects/c/libgit2/examples/lg2 diff --no-index --cached examples/file1 examples/file2
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7ebadaf in diff_print_info_init_fromdiff (pi=0x7fffffffd460, out=0x7fffffffd440, diff=0x2, format=GIT_DIFF_FORMAT_PATCH, cb=0x55555555e4de <color_printer>, 
    payload=0x7fffffffd5a0) at /home/alberto/projects/c/libgit2/src/diff_print.c:74
74		git_repository *repo = diff ? diff->repo : NULL;
(gdb) bt
#0  0x00007ffff7ebadaf in diff_print_info_init_fromdiff (pi=0x7fffffffd460, out=0x7fffffffd440, diff=0x2, format=GIT_DIFF_FORMAT_PATCH, 
    cb=0x55555555e4de <color_printer>, payload=0x7fffffffd5a0) at /home/alberto/projects/c/libgit2/src/diff_print.c:74
#1  0x00007ffff7ebc764 in git_diff_print (diff=0x2, format=GIT_DIFF_FORMAT_PATCH, print_cb=0x55555555e4de <color_printer>, payload=0x7fffffffd5a0)
    at /home/alberto/projects/c/libgit2/src/diff_print.c:690
#2  0x000055555555e1bb in lg2_diff (repo=0x555555662490, argc=5, argv=0x7fffffffd7a0) at /home/alberto/projects/c/libgit2/examples/diff.c:153
#3  0x0000555555561081 in run_command (fn=0x55555555de34 <lg2_diff>, repo=0x555555662490, args=...) at /home/alberto/projects/c/libgit2/examples/lg2.c:42
#4  0x000055555556136f in main (argc=6, argv=0x7fffffffd798) at /home/alberto/projects/c/libgit2/examples/lg2.c:110
(gdb) 
```

Didn't dig into it. Not sure what is the problem